msgpack-core/src/main/java/org/msgpack/core/buffer/ByteBufferInput.java

```diff
@@ -0,0 +1,69 @@
+//
+// MessagePack for Java
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+package org.msgpack.core.buffer;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import static org.msgpack.core.Preconditions.checkNotNull;
+
+/**
+ * {@link MessageBufferInput} adapter for {@link java.nio.ByteBuffer}
+ */
+public class ByteBufferInput
+        implements MessageBufferInput
+{
+    private ByteBuffer input;
+    private boolean isRead = false;
+
+    public ByteBufferInput(ByteBuffer input)
+    {
+        this.input = checkNotNull(input, "input ByteBuffer is null");
```

Comment:
I think it is better to call `slice()` as following to avoid this unexpected scenario:

* An user creates a new ByteBuffer and fill some data there
* The user creates a new `ByteBufferInput` with the ByteBuffer
* the user changes position of the ByteBuffer! That might be by another thread.
* ByteBufferInput is affected the change of position, unexpectedly

```
        this.input = checkNotNull(input, "input ByteBuffer is null").slice();
```